nargs() should return promargs length

JanJecmen opened the pull request.

JanJecmen committed:
nargs() should return promargs length

JanJecmen committed:
fix

o- commented:
Here are some stats for your PR:

* WARNING: the longest CI job test_release_2 took 3.75h
* are-we-fast-r suite improved by 1.02
* simple suite regressed by 0.97
* reversecomplement_naive regressed by 0.92
* fasta_naive_2 regressed by 0.93
* emptyFor regressed by 0.82
* scalarFor regressed by 0.87
* Overall benchmarks regressed by 0.99

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=2426672732&job_ids[]=2407160913&selection=all

o- commented:
Here are some stats for your PR:

* WARNING: the longest CI job test_release_2 took 2.45h
* binarytrees_2 regressed by 0.92
* spectralnorm_alt_3 regressed by 0.93
* scalarFor regressed by 0.78
* Overall benchmarks regressed by 0.99

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=2427750097&job_ids[]=2407160913&selection=all

JanJecmen committed:
Fix patching the constant pool

This should fix the bug where occasionally the close_ bytecode would find a Function in the constant pool instead of a DispatchTable.

Some cells in the constant pool are created by makeSpace with the intention that we can later patch these (used for invalidating targets after deopts).
However, the patching would also write to the C++ contents cache that maps SEXPs to pool indices, and patching multiple times would leave the old mapping outdated.

For example:
1. makeSpace 42
2. patch X into 42
3. patch Y into 42
4. X gets garbage collected
5. Z gets allocated to where X used to be
6. insert of Z finds X in contents and returns 42
Now:
- constant pool at 42 returns Y
- Z may get collected since it's not reachable from the constant pool

Also, this makes sure that we cannot patch a location that doesn't come from calling makeSpace.

o- commented:
Here are some stats for your PR:

* WARNING: the longest CI job test_features_1 took 3.43h
* simple suite improved by 1.02
* mandelbrot_naive_ascii regressed by 0.92
* pidigits regressed by 0.94
* Overall benchmarks improved by 1.01

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=2445983630&job_ids[]=2427588749&selection=all
